File: backend/app/execution_gateway/browser/wave3_visual.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class Wave3Result:
    capability_id: str
    success: bool
    duration_ms: float
    details: dict[str, Any] = field(default_factory=dict)
    error: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "capability_id": self.capability_id,
            "success": self.success,
            "duration_ms": round(self.duration_ms, 3),
            "details": dict(self.details),
            "error": self.error,
        }


def parse_payload(raw: Any) -> dict[str, Any]:
    if isinstance(raw, dict):
        return dict(raw)
    if isinstance(raw, str):
        try:
            parsed = json.loads(raw)
            return dict(parsed) if isinstance(parsed, dict) else {"text": raw}
        except json.JSONDecodeError:
            return {"text": raw}
    return {}


def detect_visual_surface(classes: str, tag_name: str = "", attrs: dict[str, str] | None = None) -> str:
    attrs = attrs or {}
    haystack = f"{tag_name} {classes} {' '.join(attrs.values())}".lower()
    if "canvas" in haystack:
        return "canvas"
    if "svg" in haystack or tag_name.lower() in {"svg", "path", "g", "circle", "rect", "line", "polyline", "polygon"}:
        return "svg"
    if "pdf" in haystack:
        return "pdf"
    if any(name in haystack for name in ("chart", "highcharts", "apexcharts", "echarts", "d3")):
        return "chart"
    if any(name in haystack for name in ("leaflet", "mapbox", "gm-style", "google-map")):
        return "map"
    if tag_name.lower() in {"video", "audio"}:
        return "media"
    return "unknown"


def execute_canvas(page: Any, locator: Any, payload: dict[str, Any]) -> Wave3Result:
    return _coordinate_action(page, locator, payload, "browser.canvas", expected_tag="canvas")


def execute_svg(page: Any, locator: Any, payload: dict[str, Any]) -> Wave3Result:
    return _coordinate_action(page, locator, payload, "browser.svg.interaction", expected_tag="svg")


def execute_chart(page: Any, locator: Any, payload: dict[str, Any]) -> Wave3Result:
    return _coordinate_action(page, locator, payload, "browser.charts.graphs")


def execute_map(page: Any, locator: Any, payload: dict[str, Any]) -> Wave3Result:
    return _coordinate_action(page, locator, payload, "browser.maps.interactive")


def execute_pdf_viewer(page: Any, locator: Any | None, payload: dict[str, Any]) -> Wave3Result:
    start = time.perf_counter()
    operation = str(payload.get("operation") or "detect")
    try:
        handle = locator.element_handle() if locator is not None else None
        state = page.evaluate(
            """([el, operation, query]) => {
              const root = el || document;
              const text = (document.body?.innerText || '').replace(/\\s+/g, ' ').trim();
              const embeds = Array.from(document.querySelectorAll('embed[type="application/pdf"], iframe[src$=".pdf"], object[type="application/pdf"], pdf-viewer'));
              const buttons = Array.from(document.querySelectorAll('button, [role="button"], input[type="search"]'));
              if (operation === 'search') {
                const search = buttons.find((node) => /search|find/i.test(node.getAttribute('aria-label') || node.getAttribute('title') || node.textContent || ''));
                if (search instanceof HTMLElement) search.click();
              }
              return {
                viewer_detected: embeds.length > 0 || location.href.toLowerCase().includes('.pdf') || text.toLowerCase().includes('page'),
                embed_count: embeds.length,
                text_match: query ? text.toLowerCase().includes(String(query).toLowerCase()) : null,
                operation,
                root_tag: root.tagName || 'document',
              };
            }""",
            [handle, operation, payload.get("query") or payload.get("text") or ""],
        )
        success = bool(state.get("viewer_detected")) and (state.get("text_match") is not False)
        return _result("browser.pdf.viewer", success, start, dict(state))
    except Exception as exc:  # noqa: BLE001
        return _result("browser.pdf.viewer", False, start, {}, str(exc)[:200])


def execute_media(page: Any, locator: Any, payload: dict[str, Any]) -> Wave3Result:
    start = time.perf_counter()
    operation = str(payload.get("operation") or "status")
    try:
        handle = locator.element_handle()
        state = page.evaluate(
            """([el, payload]) => {
              if (!(el instanceof HTMLMediaElement)) return { success: false, error: 'not_media' };
              const operation = String(payload.operation || 'status');
              if (operation === 'play') el.play?.();
              if (operation === 'pause') el.pause();
              if (operation === 'seek') el.currentTime = Number(payload.current_time ?? payload.time ?? 0);
              if (operation === 'volume') el.volume = Math.max(0, Math.min(1, Number(payload.volume ?? 1)));
              if (operation === 'fullscreen') el.requestFullscreen?.();
              return {
                success: true,
                operation,
                paused: el.paused,
                current_time: Math.round(el.currentTime * 1000) / 1000,
                volume: el.volume,
                duration: Number.isFinite(el.duration) ? el.duration : null,
              };
            }""",
            [handle, payload],
        )
        success = bool(state.get("success"))
        return _result("browser.media.controls", success, start, dict(state), None if success else str(state.get("error") or "media_failed"))
    except Exception as exc:  # noqa: BLE001
        return _result("browser.media.controls", False, start, {}, str(exc)[:200])


def execute_file_preview(page: Any, locator: Any | None, payload: dict[str, Any]) -> Wave3Result:
    start = time.perf_counter()
    try:
        handle = locator.element_handle() if locator is not None else None
        state = page.evaluate(
            """([el, expected]) => {
              const root = el || document;
              const selector = 'img, embed, object, iframe, video, audio, canvas, [role="dialog"], .preview, [data-testid*="preview"]';
              const previews = Array.from(document.querySelectorAll(selector)).filter((node) => {
                const rect = node.getBoundingClientRect?.();
                return !rect || (rect.width > 0 && rect.height > 0);
              });
              const text = (root.textContent || document.body?.innerText || '').replace(/\\s+/g, ' ').trim();
              return {
                preview_detected: previews.length > 0,
                preview_count: previews.length,
                expected_match: expected ? text.toLowerCase().includes(String(expected).toLowerCase()) : null,
                root_tag: root.tagName || 'document',
              };
            }""",
            [handle, payload.get("expected_text") or payload.get("text") or ""],
        )
        success = bool(state.get("preview_detected")) and (state.get("expected_match") is not False)
        return _result("browser.file.preview", success, start, dict(state))
    except Exception as exc:  # noqa: BLE001
        return _result("browser.file.preview", False, start, {}, str(exc)[:200])


def execute_visual_region(page: Any, locator: Any | None, payload: dict[str, Any]) -> Wave3Result:
    start = time.perf_counter()
    mode = str(payload.get("mode") or ("element" if locator is not None else "viewport"))
    try:
        if mode == "element" and locator is not None:
            image = locator.screenshot(timeout=int(payload.get("timeout_ms", 30_000)))
            details = {"mode": mode, "bytes": len(image), "target": "element"}
        elif mode == "region":
            clip = {
                "x": float(payload.get("x", 0)),
                "y": float(payload.get("y", 0)),
                "width": float(payload.get("width", 1)),
                "height": float(payload.get("height", 1)),
            }
            image = page.screenshot(clip=clip, timeout=int(payload.get("timeout_ms", 30_000)))
            details = {"mode": mode, "bytes": len(image), "clip": clip}
        else:
            image = page.screenshot(full_page=bool(payload.get("full_page", False)), timeout=int(payload.get("timeout_ms", 30_000)))
            details = {"mode": "viewport", "bytes": len(image), "full_page": bool(payload.get("full_page", False))}
        return _result("browser.visual_regions", True, start, details)
    except Exception as exc:  # noqa: BLE001
        return _result("browser.visual_regions", False, start, {"mode": mode}, str(exc)[:200])


def _coordinate_action(page: Any, locator: Any, payload: dict[str, Any], capability_id: str, expected_tag: str | None = None) -> Wave3Result:
    start = time.perf_counter()
    operation = str(payload.get("operation") or "click")
    try:
        handle = locator.element_handle()
        if handle is None:
            return _result(capability_id, False, start, {}, "target_not_found")
        box = locator.bounding_box()
        if not box:
            return _result(capability_id, False, start, {}, "target_not_visible")
        tag_name = str(page.evaluate("(el) => el.tagName ? el.tagName.toLowerCase() : ''", handle))
        if expected_tag == "canvas" and tag_name != "canvas":
            return _result(capability_id, False, start, {"tag_name": tag_name}, "canvas_not_found")
        if expected_tag == "svg" and tag_name != "svg":
            tag_name = str(page.evaluate("(el) => el.closest('svg') ? 'svg' : el.tagName.toLowerCase()", handle))
            if tag_name != "svg":
                return _result(capability_id, False, start, {"tag_name": tag_name}, "svg_not_found")
        x = float(payload.get("x", box["width"] / 2))
        y = float(payload.get("y", box["height"] / 2))
        abs_x = box["x"] + x
        abs_y = box["y"] + y
        if operation == "hover":
            page.mouse.move(abs_x, abs_y)
        elif operation == "drag":
            to_x = box["x"] + float(payload.get("to_x", x))
            to_y = box["y"] + float(payload.get("to_y", y))
            page.mouse.move(abs_x, abs_y)
            page.mouse.down()
            page.mouse.move(to_x, to_y, steps=int(payload.get("steps", 5)))
            page.mouse.up()
        elif operation == "draw":
            points = payload.get("points") if isinstance(payload.get("points"), list) else []
            page.mouse.move(abs_x, abs_y)
            page.mouse.down()
            for point in points:
                if isinstance(point, dict):
                    page.mouse.move(box["x"] + float(point.get("x", x)), box["y"] + float(point.get("y", y)))
            page.mouse.up()
        else:
            page.mouse.click(abs_x, abs_y)
        return _result(capability_id, True, start, {
            "operation": operation,
            "tag_name": tag_name,
            "x": round(x, 3),
            "y": round(y, 3),
            "absolute_x": round(abs_x, 3),
            "absolute_y": round(abs_y, 3),
            "width": round(float(box["width"]), 3),
            "height": round(float(box["height"]), 3),
        })
    except Exception as exc:  # noqa: BLE001
        return _result(capability_id, False, start, {"operation": operation}, str(exc)[:200])


def _result(capability_id: str, success: bool, start: float, details: dict[str, Any], error: str | None = None) -> Wave3Result:
    return Wave3Result(
        capability_id=capability_id,
        success=success,
        duration_ms=(time.perf_counter() - start) * 1000,
        details=details,
        error=error,
    )

File: extension/src/content/wave3_visual.ts
export interface Wave3Action {
  action_id: string
  action_type: string
  target_selector: string | null
  value: string | null
  description?: string
  reasoning?: string
  safety_level?: string
}

export interface Wave3Result {
  success: boolean
  message: string
  action_id: string
  wave3_capability?: string
  wave3_validated?: boolean
  wave3_details?: Record<string, string | number | boolean | null>
}

const WAVE3_ACTIONS = new Set([
  'canvas_action',
  'svg_action',
  'pdf_viewer',
  'chart_action',
  'map_action',
  'media_control',
  'file_preview',
  'visual_region',
])

export function parseWave3Payload(value: string | null): Record<string, unknown> {
  if (!value) return {}
  try {
    const parsed = JSON.parse(value)
    return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed as Record<string, unknown> : { text: value }
  } catch {
    return { text: value }
  }
}

export function isWave3VisualAction(actionType: string): boolean {
  return WAVE3_ACTIONS.has(actionType)
}

export async function executeWave3VisualAction(action: Wave3Action): Promise<Wave3Result | null> {
  const wave3Actions = new Set([
    'canvas_action',
    'svg_action',
    'pdf_viewer',
    'chart_action',
    'map_action',
    'media_control',
    'file_preview',
    'visual_region',
  ])
  if (!wave3Actions.has(action.action_type)) return null
  if (action.safety_level === 'danger') {
    return { success: false, message: 'Wave 3 action refused because it is marked dangerous.', action_id: action.action_id }
  }

  const payload = (() => {
    if (!action.value) return {}
    try {
      const parsed = JSON.parse(action.value)
      return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed as Record<string, unknown> : { text: action.value }
    } catch {
      return { text: action.value }
    }
  })()

  function q(selector: string | null): Element | null {
    if (!selector) return null
    try { return document.querySelector(selector) } catch { return null }
  }

  function visible(candidate: Element | null): candidate is HTMLElement | SVGElement {
    if (!(candidate instanceof HTMLElement) && !(candidate instanceof SVGElement)) return false
    const rect = candidate.getBoundingClientRect()
    const style = window.getComputedStyle(candidate)
    return rect.width > 0 && rect.height > 0 && style.display !== 'none' && style.visibility !== 'hidden'
  }

  function mouseEventAt(el: Element, type: string, x: number, y: number): void {
    const rect = el.getBoundingClientRect()
    el.dispatchEvent(new MouseEvent(type, {
      bubbles: true,
      cancelable: true,
      clientX: rect.left + x,
      clientY: rect.top + y,
      view: window,
    }))
  }

  function coordinateAction(capability: string, expected?: 'canvas' | 'svg'): Wave3Result {
    const target = q(action.target_selector)
    if (!target || !visible(target)) {
      return { success: false, message: 'Visual target not found.', action_id: action.action_id, wave3_capability: capability, wave3_validated: false }
    }
    const root = expected === 'svg' ? target.closest('svg') || target : target
    if (expected === 'canvas' && !(root instanceof HTMLCanvasElement)) {
      return { success: false, message: 'Canvas target not found.', action_id: action.action_id, wave3_capability: capability, wave3_validated: false }
    }
    if (expected === 'svg' && !(root instanceof SVGElement)) {
      return { success: false, message: 'SVG target not found.', action_id: action.action_id, wave3_capability: capability, wave3_validated: false }
    }
    const rect = root.getBoundingClientRect()
    const x = Number(payload.x ?? rect.width / 2)
    const y = Number(payload.y ?? rect.height / 2)
    const operation = String(payload.operation ?? 'click')
    if (operation === 'hover') mouseEventAt(root, 'mousemove', x, y)
    else if (operation === 'drag') {
      mouseEventAt(root, 'mousedown', x, y)
      mouseEventAt(root, 'mousemove', Number(payload.to_x ?? x), Number(payload.to_y ?? y))
      mouseEventAt(root, 'mouseup', Number(payload.to_x ?? x), Number(payload.to_y ?? y))
    } else if (operation === 'draw') {
      mouseEventAt(root, 'mousedown', x, y)
      const points = Array.isArray(payload.points) ? payload.points : []
      for (const point of points) {
        if (point && typeof point === 'object') {
          const p = point as Record<string, unknown>
          mouseEventAt(root, 'mousemove', Number(p.x ?? x), Number(p.y ?? y))
        }
      }
      mouseEventAt(root, 'mouseup', x, y)
    } else {
      mouseEventAt(root, 'click', x, y)
    }
    return {
      success: true,
      message: `${capability} ${operation} completed.`,
      action_id: action.action_id,
      wave3_capability: capability,
      wave3_validated: true,
      wave3_details: { operation, x, y, width: Math.round(rect.width), height: Math.round(rect.height) },
    }
  }

  function pdfViewer(): Wave3Result {
    const text = (document.body?.innerText || '').replace(/\s+/g, ' ').trim()
    const embeds = document.querySelectorAll('embed[type="application/pdf"], iframe[src$=".pdf"], object[type="application/pdf"], pdf-viewer')
    const query = String(payload.query ?? payload.text ?? '')
    const detected = embeds.length > 0 || location.href.toLowerCase().includes('.pdf') || text.toLowerCase().includes('page')
    const matched = query ? text.toLowerCase().includes(query.toLowerCase()) : true
    return { success: detected && matched, message: detected ? 'PDF viewer detected.' : 'PDF viewer not detected.', action_id: action.action_id, wave3_capability: 'browser.pdf.viewer', wave3_validated: detected && matched, wave3_details: { embed_count: embeds.length, query_match: matched } }
  }

  function mediaControl(): Wave3Result {
    const target = q(action.target_selector)
    if (!(target instanceof HTMLMediaElement)) return { success: false, message: 'Media element not found.', action_id: action.action_id, wave3_capability: 'browser.media.controls', wave3_validated: false }
    const operation = String(payload.operation ?? 'status')
    if (operation === 'play') void target.play?.()
    if (operation === 'pause') target.pause()
    if (operation === 'seek') target.currentTime = Number(payload.current_time ?? payload.time ?? 0)
    if (operation === 'volume') target.volume = Math.max(0, Math.min(1, Number(payload.volume ?? 1)))
    if (operation === 'fullscreen') void target.requestFullscreen?.()
    return { success: true, message: `Media ${operation} completed.`, action_id: action.action_id, wave3_capability: 'browser.media.controls', wave3_validated: true, wave3_details: { operation, paused: target.paused, current_time: Math.round(target.currentTime * 1000) / 1000, volume: target.volume } }
  }

  function filePreview(): Wave3Result {
    const root = q(action.target_selector) || document.body
    const previews = Array.from(document.querySelectorAll('img, embed, object, iframe, video, audio, canvas, [role="dialog"], .preview, [data-testid*="preview"]')).filter(visible)
    const expected = String(payload.expected_text ?? payload.text ?? '')
    const text = (root?.textContent || '').replace(/\s+/g, ' ').trim()
    const matched = expected ? text.toLowerCase().includes(expected.toLowerCase()) : true
    const detected = previews.length > 0
    return { success: detected && matched, message: detected ? 'File preview detected.' : 'File preview not detected.', action_id: action.action_id, wave3_capability: 'browser.file.preview', wave3_validated: detected && matched, wave3_details: { preview_count: previews.length, expected_match: matched } }
  }

  function visualRegion(): Wave3Result {
    const target = q(action.target_selector)
    const mode = String(payload.mode ?? (target ? 'element' : 'viewport'))
    const rect = target?.getBoundingClientRect()
    return {
      success: true,
      message: 'Visual capture metadata prepared.',
      action_id: action.action_id,
      wave3_capability: 'browser.visual_regions',
      wave3_validated: true,
      wave3_details: {
        mode,
        x: Number(payload.x ?? rect?.left ?? 0),
        y: Number(payload.y ?? rect?.top ?? 0),
        width: Number(payload.width ?? rect?.width ?? window.innerWidth),
        height: Number(payload.height ?? rect?.height ?? window.innerHeight),
      },
    }
  }

  if (action.action_type === 'canvas_action') return coordinateAction('browser.canvas', 'canvas')
  if (action.action_type === 'svg_action') return coordinateAction('browser.svg.interaction', 'svg')
  if (action.action_type === 'chart_action') return coordinateAction('browser.charts.graphs')
  if (action.action_type === 'map_action') return coordinateAction('browser.maps.interactive')
  if (action.action_type === 'pdf_viewer') return pdfViewer()
  if (action.action_type === 'media_control') return mediaControl()
  if (action.action_type === 'file_preview') return filePreview()
  if (action.action_type === 'visual_region') return visualRegion()
  return null
}
